File: csharp/src/DarkStrata.CredentialCheck/Types/ClientOptions.cs
using System;
using System.Diagnostics.CodeAnalysis;

namespace DarkStrata.CredentialCheck;

/// <summary>
/// Configuration options for the DarkStrata credential check client.
/// </summary>
public sealed class ClientOptions
{
    public ClientOptions()
    {
    }

    [SetsRequiredMembers]
    public ClientOptions(string apiKey, string? baseUrl = null, TimeSpan? timeout = null, int? retries = null, bool? enableCaching = null, TimeSpan? cacheTtl = null)
    {
        ApiKey = apiKey;
        BaseUrl = baseUrl;
        Timeout = timeout;
        Retries = retries;
        EnableCaching = enableCaching;
        CacheTtl = cacheTtl;
    }

    /// <summary>
    /// Your DarkStrata API key (JWT token).
    /// Obtain this from your DarkStrata dashboard.
    /// </summary>
    public required string ApiKey { get; init; }

    /// <summary>
    /// Base URL for the DarkStrata API.
    /// </summary>
    /// <remarks>Default: https://api.darkstrata.io/v1/</remarks>
    public string? BaseUrl { get; init; }

    /// <summary>
    /// Request timeout.
    /// </summary>
    /// <remarks>Default: 30 seconds</remarks>
    public TimeSpan? Timeout { get; init; }

    /// <summary>
    /// Number of retry attempts for failed requests.
    /// </summary>
    /// <remarks>Default: 3</remarks>
    public int? Retries { get; init; }

    /// <summary>
    /// Enable in-memory caching of API responses.
    /// Cache is automatically invalidated when the server time window changes.
    /// </summary>
    /// <remarks>Default: true</remarks>
    public bool? EnableCaching { get; init; }

    /// <summary>
    /// Cache time-to-live.
    /// Should align with server time window (1 hour).
    /// </summary>
    /// <remarks>Default: 1 hour</remarks>
    public TimeSpan? CacheTtl { get; init; }
}

File: sdks/java/src/main/java/io/darkstrata/credentialcheck/ClientOptions.java
package io.darkstrata.credentialcheck;

/**
 * Configuration options for the DarkStrata client.
 */
public class ClientOptions {

    private final String apiKey;
    private final String baseUrl;
    private final long timeout;
    private final int retries;
    private final boolean enableCaching;
    private final long cacheTTL;

    private ClientOptions(Builder builder) {
        this.apiKey = builder.apiKey;
        this.baseUrl = builder.baseUrl;
        this.timeout = builder.timeout;
        this.retries = builder.retries;
        this.enableCaching = builder.enableCaching;
        this.cacheTTL = builder.cacheTTL;
    }

    public String getApiKey() {
        return apiKey;
    }

    public String getBaseUrl() {
        return baseUrl;
    }

    public long getTimeout() {
        return timeout;
    }

    public int getRetries() {
        return retries;
    }

    public boolean isEnableCaching() {
        return enableCaching;
    }

    public long getCacheTTL() {
        return cacheTTL;
    }

    /**
     * Create a new builder with the required API key.
     */
    public static Builder builder(String apiKey) {
        return new Builder(apiKey);
    }

    public static class Builder {
        private final String apiKey;
        private String baseUrl = Constants.DEFAULT_BASE_URL;
        private long timeout = Constants.DEFAULT_TIMEOUT;
        private int retries = Constants.DEFAULT_RETRIES;
        private boolean enableCaching = true;
        private long cacheTTL = Constants.DEFAULT_CACHE_TTL;

        private Builder(String apiKey) {
            this.apiKey = apiKey;
        }

        /**
         * Set the base URL for the API (default: https://api.darkstrata.io/v1/).
         */
        public Builder baseUrl(String baseUrl) {
            this.baseUrl = baseUrl;
            return this;
        }

        /**
         * Set the request timeout in milliseconds (default: 30000).
         */
        public Builder timeout(long timeout) {
            this.timeout = timeout;
            return this;
        }

        /**
         * Set the number of retry attempts (default: 3).
         */
        public Builder retries(int retries) {
            this.retries = retries;
            return this;
        }

        /**
         * Enable or disable caching (default: true).
         */
        public Builder enableCaching(boolean enableCaching) {
            this.enableCaching = enableCaching;
            return this;
        }

        /**
         * Set the cache TTL in milliseconds (default: 3600000 = 1 hour).
         */
        public Builder cacheTTL(long cacheTTL) {
            this.cacheTTL = cacheTTL;
            return this;
        }

        public ClientOptions build() {
            return new ClientOptions(this);
        }
    }
}
